library/Vanilla/Theme/ThemeCache.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * @author Adam Charron <adam.c@vanillaforums.com>
+ * @copyright 2009-2020 Vanilla Forums Inc.
+ * @license GPL-2.0-only
+ */
+
+namespace Vanilla\Theme;
+
+/**
+ * Implement caching for the theme service.
+ */
+class ThemeCache {
+
+    /** @var string Cache key for holding other cache keys for easy invalidation. */
+    const CACHE_HOLDER_KEY = 'singleThemeCacheHolder';
+
+    /** @var string */
+    const CACHE_KEY = "theme";
+
+    /** @var int 10 minute cache interval. */
+    const CACHE_TTL = 600;
+
+    /** @var \Gdn_Cache */
+    private $cache;
+
+    /**
+     * DI.
+     *
+     * @param \Gdn_Cache $cache
+     */
+    public function __construct(\Gdn_Cache $cache) {
+        $this->cache = $cache;
+    }
+
+    /**
+     * Try to fetch a theme from the cache.
+     *
+     * @param string $key
+     * @return array|null
+     */
+    public function get(string $key): ?Theme {
+        $cacheResult = $this->cache->get($key);
+        if ($cacheResult) {
+            $theme = unserialize($cacheResult) ?: null;
+            if ($theme instanceof Theme) {
+                $theme->setIsCacheHit(true);
+            }
+            return $theme;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Cache a value.
+     *
+     * @param string $key
+     * @param Theme $theme
+     */
+    public function set(string $key, Theme $theme) {
+        $this->addCacheKey($key);
+        $this->cache->store($key, serialize($theme), [
+            \Gdn_Cache::FEATURE_EXPIRY => self::CACHE_TTL,
+        ]);
+    }
+
+    /**
+     * Clear all known caches of navigation.
+     */
+    public function clear() {
+        $allCacheKeys = $this->cache->get(self::CACHE_HOLDER_KEY) ?: [];
+
+        foreach ($allCacheKeys as $cacheKey) {
+            $result = $this->cache->remove($cacheKey);
+            if ($result === \Gdn_Cache::CACHEOP_FAILURE) {
+                trigger_error('Failed to clear theme cache', E_USER_NOTICE);
+            }
+        }
+        $this->cache->remove(self::CACHE_HOLDER_KEY);
+    }
+
+    /**
+     * Generate a cache key.
+     *
+     * @param string|int $themeID
+     * @param array $args
+     * @return string
+     */
+    public function cacheKey($themeID, array $args = []): string {
+        // Put args in a stable order.
+        ksort($args);
+
+        $cacheKey = self::CACHE_KEY . '_' . $themeID;
+        foreach ($args as $key => $val) {
+            $cacheKey .= '_' . $key . '-' . json_encode($val);
+        }
+        return $cacheKey;
```

Comment:
This code will generate huge `$cacheKey`.

To make unique key we can just generate 
```suggestion
        ksort($args);

        $cacheKey = self::CACHE_KEY . '_' . $themeID.md5(json_encode($args));
        
        return $cacheKey;
```